Showing a problem when using ObjectController with ValidatorViewSupport

vitch opened the pull request:
These commits show the problem that I was describing in issue #3. It turns out I was wrong about the cause (no problem with the latest version of ember or with using ember-data and ember-validation). The issue with  `ValidatorViewSupport` happens when you are using it along with an `ObjectController` as you can see if you try to use example 2 in my version of the code...

Is this something that can be worked around or should I bind the properties as `model.propertyName` rather than just `propertyName`?

vitch committed:
Upgraded ember to v1.3.1 and changed deprecated properties...

vitch committed:
Using ember-data for the model to make sure this doesn't cause
problems..

vitch committed:
Switched to using the {{input}} for plain text fields and custom
components for the subclasses...

vitch committed:
Using the default model name for the model and an ObjectController as per
the ember getting started guides...

Dani723 commented:
Hi,

thanks for reporting the problem and updating the demo :).
I think the problem lies within the use of the new Component class. The component is completely isolated from the outer controller. The ValidatorViewSupport mixin tries to obtain the context of the component, which is not the ObjectController. Therefore it has not the "validate" implementation. 

This lib was written, when there was no Component class.
But as components are surely the right choice for this use-case, ember-validation urgently needs an update. I try to get it done within the next 2 or 3 days.

vitch committed:
Hacky fix to allow ValidatorViewSupport views to access the validator
hosted by an ObjectController.

I'm not sure if this approach is correct but it seems to make the example
work...

vitch commented:
I suspected it was the component class but I now don't think it is... I think that the problem is related to using an ObjectController and referring to properties which get proxied through to the model. If you step through my commits you'll see that everything was still working fine after a246945 when I introduced components... 

I've just pushed a hacky "fix" which seems to make your second example work correctly even when using an ObjectController. I'm not sure if the fix is "correct" though. And in my application there still seems to be problems even with this fix in place...

Thanks for looking into it!

Dani723 commented:
Hi, you are absolutely right.
The value bindings in the templates are proxied. The ValidatorViewMixin grabs the object controller instead of the model, which doesn't implement the ValidatorMixin. 
It works without the fix, when the the binding is set to the model instead of the controller.

{{input id="inputName" value=model.name}}

Your fix seems legit to me, as the ObjectController should be the only exception from the default behavior.  I've extended your code by an instance detection.

Thanks for your contibution :).
